### DISCUSSION TITLE
Dispatch before transforming

### DISCUSSION THREAD

**Discussion by ta-le** [reduxjs/redux-toolkit#3981] [Q&A]
I have a query where the fetch response is transformed using `transformResponse`. 
I also want to search the original (not the transformed) response for some data to add to another Redux slice (meaning I need the dispatch API). 

Since `dispatch` is not available from `transformResponse`, is there another place where I have access to the original response + dispatch?

Thanks in advance!

**Comment by EskiMojo14** [comment]
in your other slice, you can respond to the original endpoint fulfilled action, which will have the response instance attached.
```ts
const slice = createSlice({
  ...config,
  extraReducers(builder) {
    builder.addMatcher(api.endpoints.getPost.matchFulfilled, (state, action) => {
      const { response } = action.meta.baseQueryMeta
      try {
        const json = response.json();
      } catch {}
    })
  }
})
```
alternatively if you wanted to do this in a place where you have access to dispatch, you could use a [listener](https://redux-toolkit.js.org/api/createListenerMiddleware).

**Reply by ta-le** [reply]
Thanks, that's exactly what I was looking for! 

However, I'm now running into the problem that `json()` actually needs to be awaited for. From my understanding, it's not allowed to run async code in here, is it?

**Reply by EskiMojo14** [reply]
hmm, you're right.

honestly i'd probably do something like 

```ts
transformResponse: (response) => ({ original: response, transformed: transform(response) })

const slice = createSlice({
  ...config,
  extraReducers(builder) {
    builder.addMatcher(api.endpoints.getPost.matchFulfilled, (state, action) => {
      const { original } = action.payload
    })
  }
})
```

it does mean that you'd have to deal with selecting out transformed when you call the hook, but I think this is still the simplest way.

**Reply by ta-le** [reply]
Yes that's always a solid plan B!

With the solution above, running `reponse.json()` also runs into `TypeError: Already read` error, the body is already used by RTK Query.
However, in the `action.meta.baseQueryMeta.response` data there are `_bodyInit` and `_bodyText`. They contain what you would expect when doing `response.text()` as a string (the unparsed JSON data). 

Is it wise to use those?

**Reply by phryneas** [reply]
I fear not - those are probably implementation details of your runtime or a polyfill you are using.
